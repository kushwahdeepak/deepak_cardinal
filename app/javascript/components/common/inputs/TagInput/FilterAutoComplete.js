import React, { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import axios from 'axios'
import './style/FilterAutocomplete.scss'

const FilterAutocomplete = ({ getFilterValue, testAttr, onBlur }) => {
    const [searchLoc, setSearchLoc] = useState('')
    const [searchValue, setSearchValue] = useState('')
    const [state, setState] = useState({
        showSuggestions: false,
        suggestions: [],
        selectedSearchValue: '',
    })
    const selectRef = React.createRef()
    const { showSuggestions, suggestions, selectedSearchValue } = state

    const getSearchValue = async (searchTerm) => {
        let url
        if (testAttr == 'company_names') {
            url = '/filter_candidate_on_company'
        } else if (testAttr == 'schools'){
            url = '/filter_candidate_on_education'
        } else if (testAttr == 'city'){
            url = '/filter_candidate_on_location?flag=true'
        } else if (testAttr === 'companyNames'){
            url = `/organizations/get_organization?search=${searchTerm}`
        } else {
            url = '/filter_candidate_on_location'
        }
        const CSRF_Token = document
            .querySelector('meta[name="csrf-token"]')
            .getAttribute('content')

        try {
            let response
            if (testAttr === 'companyNames') {
                 response = await axios.get(url)
                console.log(response)
                let organizations = response.data.organizations
                console.log(organizations)
                setState((prevState) => ({
                    ...prevState,
                    suggestions: [...organizations],
                    showSuggestions: true,
                }))
            } else {
                 response = await axios.post(
                    url,
                    { filter_word: searchTerm },
                    {
                        headers: {
                            'content-type': 'application/json',
                            'X-CSRF-Token': CSRF_Token,
                        },
                    }
                )
                setState((prevState) => ({
                    ...prevState,
                    suggestions: response.data.filter,
                    showSuggestions: true,
                }))
            }
        } catch (e) {
            console.log(e.message, 'error')
        }
    }

    const handleSelectLocation = (e) => {
        setSearchValue(e.target.value)
    }

    const handleOnBlur = (e) => {
        if(!e.target.value){
            onBlur()
        }
    }

    const handleFileterValue = (testAttr, suggestion) => {
        switch(testAttr){
            case 'company_names':
               return suggestion.company_name
            case 'state':   
                return suggestion.state
            case 'city':   
                 return suggestion.city +
                ', ' +
                suggestion.state +
                ', ' +
                suggestion.country
            case 'schools':
                return suggestion.name
            case 'companyNames':
                return suggestion.name
            default: 
                break     
        }

    }

    useEffect(() => {
        getFilterValue(selectedSearchValue)
        setSearchValue('')
    }, [selectedSearchValue])

    let suggestionsListComponent
    if (showSuggestions && searchLoc) {
        if (suggestions.length) {
            suggestionsListComponent = (
                <select
                    ref={selectRef}
                    className="suggestions"
                    value={searchValue}
                    size={
                        suggestions.length < 5
                            ? suggestions.length == 1
                                ? 2
                                : suggestions.length
                            : 5
                    }
                    style={{width:`${testAttr === `companyNames` ? '134px' : '160px' }`}}
                    onChange={(e) => {
                        handleSelectLocation(e)
                    }}
                    onKeyDown={(e) => {
                        if (e.key === 'Enter') {
                            setSearchValue(e.target.value)
                            setSearchLoc(
                                e.target.options[e.target.selectedIndex].text
                            )
                            setState({
                                showSuggestions: false,
                                suggestions: [],
                                selectedSearchValue: e.target.value,
                            })
                        }
                    }}
                >
                    {suggestions.map((suggestion, index) => {
                        return (
                            <option
                                key={suggestion.id}
                                onClick={(e) => {
                                    setSearchValue(suggestion.id)
                                    setSearchLoc(handleFileterValue(testAttr, suggestion))
                                    setState({
                                        showSuggestions: false,
                                        suggestions: [],
                                        selectedSearchValue: suggestion,
                                    })
                                }}
                                value={suggestion.id}
                                style={{ cursor: 'pointer' }}
                            >
                                {handleFileterValue(testAttr, suggestion)}
                            </option>
                        )
                    })}
                </select>
            )
        } else {
            suggestionsListComponent = (
               <div className="no-suggestions" />
            )
        }
    }

    return (
        <div className="wrapper">
            <Row style={{ margin: '0px' }}>
                <Col style={{ padding: '0', margin: '0px' }}>
                    <input
                        className="input search_text"
                        type="text"
                        id="location_search"
                        autoComplete="off"
                        onBlur={(e)=> handleOnBlur(e)}
                        onChange={(e) => {
                            getSearchValue(e.target.value)
                            setSearchLoc(e.target.value)
                        }}
                        placeholder='Search...'
                        onKeyDown={(e) => {
                            if (e.key === 'ArrowDown') {
                                if (selectRef.current) {
                                    selectRef.current.focus()
                                }
                            }
                            if (e.key === 'Enter') {
                                e.preventDefault()
                                e.stopPropagation()
                                if (selectRef.current) {
                                    const selectedOption =
                                        selectRef.current.options[0]
                                    setSearchValue(selectedOption.value)
                                    setSearchLoc(selectedOption.text)
                                    setState({
                                        showSuggestions: false,
                                        suggestions: [],
                                        selectedSearchValue: selectedOption.value,
                                    })
                                }
                            }
                        }}
                        value={searchLoc }
                        autoFocus
                    />
                    {suggestionsListComponent}
                </Col>
            </Row>
        </div>
    )
}

FilterAutocomplete.propTypes = {
    getFilterValue: PropTypes.func.isRequired,
}

export default FilterAutocomplete
