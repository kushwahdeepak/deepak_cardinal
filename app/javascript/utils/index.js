import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
