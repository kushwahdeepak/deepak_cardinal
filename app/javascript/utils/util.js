import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
