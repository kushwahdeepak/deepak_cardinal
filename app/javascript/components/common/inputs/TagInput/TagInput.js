import React, { useEffect, useState, useRef } from 'react'
import TagList from '../../TagList/TagList'
import Button from 'react-bootstrap/Button'
import { nanoid } from 'nanoid'
import isNil from 'lodash.isnil'

import './style/tagInput.scss'
import FilterAutoComplete from './FilterAutoComplete'
import Util from '../../../../utils/util'

function TagInput({
    label,
    setInputState,
    setFilter,
    testFunc,
    testAttr,
    initialValues = [],
    candidatePage,
    showTag=true
}) {
    const [tags, setTags] = useState([...initialValues])
    const [showInput, setShowInput] = useState(false)
    const inputRef = useRef()
    const buttonRef = useRef()
    
    useEffect(() => {
        if (!isNil(setFilter)) {
            console.assert(
                !(isNil(testFunc) && isNil(testAttr)),
                'TagInput requires either `testFunc` or `testAttr` to be provided'
            )
            testFunc = testFunc ?? getDefaultTestFunc(testAttr)
            
            setFilter(
                testAttr,
                tags.length === 0
                    ? null
                    : (candidate) => testFunc(candidate, tags)
            )
        }

        if (!isNil(setInputState)) {
            setInputState(
                testAttr,
                tags.map((t) => t.value)
            )
        }

        if (buttonRef.current) {
            buttonRef.current.focus()
        }
    }, [tags])

    useEffect(() => {
        if (inputRef.current) {
            inputRef.current.focus()
        }
    })

    useEffect(() => {
        setTags(initialValues)
    }, [initialValues.length])

    const handleInputKey = (e) => {
        if (e.which === 13 && e.target.value != ''){
            setTags(tags.concat({ id: nanoid(), value: e.target.value }))
            setShowInput(false)
        } else if (e.which === 13 && e.target.value === '') {
            setShowInput(false)
        }
    }

    const handleSelectName = (e, name) => {
        let val = ''
        if(name == "state") {
          val = e.state
        } else if(name == 'city') {
          val = (e.city+', '+e.state+', '+e.country)
        }
        if (e != '' && (e.company_name != undefined || (val && e.state != undefined|| e.city != undefined) || e.name !== undefined)){
            setTags(tags.concat({ id: nanoid(), value: (e.company_name || val || e.name)}))
            setShowInput(false)
        }
    }


    return (
        <div>
            {showTag &&
            <TagList
                tags={tags}
                removeTagFunc={(id) =>
                    setTags(tags.filter((tag) => tag.id !== id))
                }
                maxTagWidth={'100%'}
                shadows
            ></TagList>
            }
            {!showInput && (
                <Button
                    variant="grey"
                    ref={buttonRef}
                    onClick={() => {
                        showTag && setShowInput(true)
                    }}
                    style={{backgroundColor: candidatePage ? '#E6E9FC' : '' }}
                    className={`${candidatePage ? "btn btn-info p-1 rounded-circle btn-sm" : ''} ${showTag ? "" : "pointer-block"}` }
                >
                    {plusIcon()}
                </Button>
            )}
            {showInput &&  testAttr !== 'company_names' &&  testAttr !== 'companyNames' && testAttr !== 'city' && testAttr !== 'state'  && testAttr !== 'schools' && (
                    <div style={{ width: '100%' }}>
                        <input
                            type="text"
                            placeholder='Search...'
                            className="tagTextInput"
                            onKeyPress={handleInputKey}
                            onBlur={() => setShowInput(false)}
                            ref={inputRef}
                        />
                    </div>
                )}
            {showInput &&  (testAttr == 'company_names' || testAttr === 'companyNames' || testAttr == 'city' || testAttr == 'state' || testAttr == 'schools') && (
                    <FilterAutoComplete
                        getFilterValue={(e) => {
                            handleSelectName(e, testAttr)
                        }}
                        testAttr={testAttr}
                        onBlur={() => {setShowInput(false)}}
                    />
                )}
        </div>
    )
}

function getDefaultTestFunc(testAttr) {
    return (candidate, tags) => {
        return Util.attributePassesStringSetFilter(
            candidate,
            testAttr,
            tags.map((t) => t.value)
        )
    }
}

function plusIcon() {
    return (
        <svg
            width="27"
            height="27"
            fill="none"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            className="css-i6dzq1"
            viewBox="0 0 24 24"
            style={{marginRight: '1px' }}
        >
            <path d="M12 5L12 19"></path>
            <path d="M5 12L19 12"></path>
        </svg>
    )
}

export default TagInput
