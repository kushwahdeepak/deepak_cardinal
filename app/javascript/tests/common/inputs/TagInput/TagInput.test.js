import React from 'react';
import TagInput from '../../../../components/common/inputs/TagInput/TagInput';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr } from '../../../../utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<TagInput {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<TagInput {...props}/>);
};


describe('TagInput component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {

		const props = {
		};
		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	afterEach(() => {
		fullWrapper.unmount();
	});

	it('should render TagInput component', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it('should match with a snapshot', () => {
		const component = renderer.create(<TagInput />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('TagInput component with props', () => {

	});
	describe('TagInput component without props', () => {

	});
});
