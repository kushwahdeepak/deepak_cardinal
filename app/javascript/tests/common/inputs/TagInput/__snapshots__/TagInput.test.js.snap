// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagInput component should match with a snapshot 1`] = `
<div
  className="row"
>
  <button
    className="btn btn-white"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <svg
      className="css-i6dzq1"
      fill="none"
      height="24"
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width="24"
    >
      <path
        d="M12 5L12 19"
      />
      <path
        d="M5 12L19 12"
      />
    </svg>
  </button>
</div>
`;
